service/discovery/openstack/account_discover.go

```diff
@@ -26,24 +26,63 @@
 package openstack
 
 import (
-	"clouditor.io/clouditor/v2/api/ontology"
+	"fmt"
 
+	"clouditor.io/clouditor/v2/api/discovery"
+	"clouditor.io/clouditor/v2/api/ontology"
 	"github.com/gophercloud/gophercloud/v2/openstack/identity/v3/domains"
 	"github.com/gophercloud/gophercloud/v2/openstack/identity/v3/projects"
 )
 
+// discoverDomains discovers domains.
+func (d *openstackDiscovery) discoverDomains() (list []ontology.IsResource, err error) {
+	var opts domains.ListOptsBuilder = &domains.ListOpts{}
+	list, err = genericList(d, d.identityClient, domains.List, d.handleDomain, domains.ExtractDomains, opts)
+
+	if err != nil {
+		// if we cannot retrieve the domain information by calling the API or from the environment variables, we will add the information manually if we already got the domain ID and/or domain name
+		log.Debugf("could not discover domains due to insufficient permissions, but we can proceed with less domain information: %v", err)
+
+		if d.domain.domainID == "" && d.domain.domainName == "" {
+			err := fmt.Errorf("neither the domain ID nor the domain name are available: %v", err)
+			return nil, err
+		}
+
+		r := &ontology.Account{
+			Id:   d.domain.domainID,
+			Name: d.domain.domainName,
+			Raw:  discovery.Raw("Domain information manually added."),
+		}
+
+		list = append(list, r)
+	}
+
+	return list, nil
+}
+
 // discoverProjects discovers projects/tenants. OpenStack project and tenant are interchangeable.
 func (d *openstackDiscovery) discoverProjects() (list []ontology.IsResource, err error) {
 	var opts projects.ListOptsBuilder = &projects.ListOpts{}
 	list, err = genericList(d, d.identityClient, projects.List, d.handleProject, projects.ExtractProjects, opts)
 
-	return
-}
+	if err != nil {
+		// if we cannot retrieve the project information by calling the API or from the environment variables, we will add the information manually if we already got the project ID and/or project name
+		log.Debugf("could not discover projects/tenants due to insufficient permissions, but we can proceed with less project/tenant information: %v", err)
 
-// discoverDomains discovers domains
-func (d *openstackDiscovery) discoverDomains() (list []ontology.IsResource, err error) {
-	var opts domains.ListOptsBuilder = &domains.ListOpts{}
-	list, err = genericList(d, d.identityClient, domains.List, d.handleDomain, domains.ExtractDomains, opts)
+		if d.project.projectID == "" && d.project.projectName == "" {
+			err := fmt.Errorf("neither the project ID nor the project name are available: %v", err)
+			return nil, err
+		}
+
+		r := &ontology.ResourceGroup{
+			Id:       d.project.projectID,
+			Name:     d.project.projectName,
+			ParentId: &d.domain.domainID,
```

Comment:
Is it guaranteed that domainID is not empty (if it is important)?

Reply:
The domain ID must be set by the env variables or must be discovered by the domains. If the discovery fails it should be set via the env variables. If both does not happen the error occurs.